builtin/logical/pki/path_config_ca.go

```diff
@@ -13,24 +16,443 @@ func pathConfigCA(b *backend) *framework.Path {
+		HelpDescription: pathConfigCAGenerateHelpDesc,
+	}
+
+	for k, v := range generateSchema {
+		ret.Fields[k] = v
+	}
+
+	ret.Fields["max_path_length"] = &framework.FieldSchema{
+		Type:        framework.TypeInt,
+		Default:     -1,
+		Description: "The maximum allowable path length",
+	}
+
+	return ret
+}
+
+func pathGenerateIntermediateCA(b *backend) *framework.Path {
+	ret := &framework.Path{
+		Pattern: "config/ca/generate/intermediate/" + framework.GenericNameRegex("exported"),
+
+		Fields: rootAndSignSchema,
+
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.WriteOperation: b.pathCAGenerateIntermediate,
+		},
+
+		HelpSynopsis:    pathConfigCAGenerateHelpSyn,
+		HelpDescription: pathConfigCAGenerateHelpDesc,
+	}
+
+	for k, v := range generateSchema {
+		ret.Fields[k] = v
+	}
+
+	return ret
+}
+
+func pathSignIntermediateCA(b *backend) *framework.Path {
+	ret := &framework.Path{
+		Pattern: "config/ca/sign",
+
+		Fields: rootAndSignSchema,
+
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.WriteOperation: b.pathCASignIntermediate,
+		},
+
+		HelpSynopsis:    pathConfigCASignHelpSyn,
+		HelpDescription: pathConfigCASignHelpDesc,
+	}
+
+	ret.Fields["csr"] = &framework.FieldSchema{
+		Type:        framework.TypeString,
+		Default:     "",
+		Description: `PEM-format CSR to be signed.`,
+	}
+
+	ret.Fields["use_csr_values"] = &framework.FieldSchema{
+		Type:    framework.TypeBool,
+		Default: false,
+		Description: `If true, then:
+1) Subject information, including names and alternate
+names, will be preserved from the CSR rather than
+using values provided in the other parameters to
+this path;
+2) Any key usages requested in the CSR will be
+added to the basic set of key usages used for CA
+certs signed by this path; for instance,
+the non-repudiation flag.`,
+	}
+
+	ret.Fields["max_path_length"] = &framework.FieldSchema{
+		Type:        framework.TypeInt,
+		Default:     -1,
+		Description: "The maximum allowable path length",
+	}
+
+	return ret
+}
+
+func (b *backend) getGenerationParams(
+	data *framework.FieldData,
+) (exported bool, format string, role *roleEntry, errorResp *logical.Response) {
+	exportedStr := data.Get("exported").(string)
+	switch exportedStr {
+	case "exported":
+		exported = true
+	case "internal":
+	default:
+		errorResp = logical.ErrorResponse(
+			`The "exported" path parameter must be "internal" or "exported"`)
+		return
+	}
+
+	format = data.Get("format").(string)
+	switch format {
+	case "pem":
+	case "der":
+	default:
+		errorResp = logical.ErrorResponse(
+			`The "format" path parameter must be "pem" or "der"`)
+		return
+	}
+
+	role = &roleEntry{
+		TTL:              data.Get("ttl").(string),
+		KeyType:          data.Get("key_type").(string),
+		KeyBits:          data.Get("key_bits").(int),
+		AllowLocalhost:   true,
+		AllowAnyName:     true,
+		AllowIPSANs:      true,
+		EnforceHostnames: false,
+	}
+
+	switch role.KeyType {
```

Comment:
I've seen this block in a few places, can we have a shared `ValidateKeyLength(algorithm, bits)` function instead